com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainMethodHandler.java

```diff
@@ -0,0 +1,162 @@
+            if (method != null) {
+                result.add(method);
+            }
+        }
+
+        return result;
+    }
+
+    private static IMethod getMainMethod(IType type) throws JavaModelException {
+        for (IMethod method : type.getMethods()) {
+            // Have at most one main method in the member methods of the type.
+            if (method.isMainMethod()) {
+                return method;
+            }
+        }
+
+        return null;
+    }
+
+    private static List<IType> getPotentialMainClassTypes(ICompilationUnit compilationUnit) throws JavaModelException {
+        List<IType> result = new ArrayList<>();
+        IType[] topLevelTypes = compilationUnit.getTypes();
+        result.addAll(Arrays.asList(topLevelTypes));
+        for (IType type : topLevelTypes) {
+            result.addAll(getPotentialMainClassTypesInChildren(type));
+        }
+
+        return result;
+    }
+
+    private static List<IType> getPotentialMainClassTypesInChildren(IType type) throws JavaModelException {
+        IType[] children = type.getTypes();
+        if (children.length == 0) {
+            return Collections.emptyList();
+        }
+
+        List<IType> result = new ArrayList<>();
+        for (IType child : children) {
+            // main method can only exist in the static class or top level class.
+            if (child.isClass() && Flags.isStatic(child.getFlags())) {
+                result.add(child);
+                result.addAll(getPotentialMainClassTypesInChildren(child));
+            }
+        }
+
+        return result;
+    }
+
+    private static MainMethod constructMainMethodCodeLens(ICompilationUnit typeRoot, IMethod method) throws JavaModelException {
+        final Range range = getRange(typeRoot, method);
+        IResource resource = typeRoot.getResource();
+        if (resource != null) {
+            IProject project = resource.getProject();
+            if (project != null) {
+                String mainClass = method.getDeclaringType().getFullyQualifiedName();
+                IJavaProject javaProject = JdtUtils.getJavaProject(project);
+                if (javaProject != null) {
+                    String moduleName = JdtUtils.getModuleName(javaProject);
+                    if (moduleName != null) {
+                        mainClass = moduleName + "/" + mainClass;
+                    }
+                }
+
+                String projectName = ProjectsManager.DEFAULT_PROJECT_NAME.equals(project.getName()) ? null : project.getName();
+                return new MainMethod(range, mainClass, projectName);
+            }
+        }
+
+        return null;
+    }
+
+    private static Range getRange(ICompilationUnit typeRoot, IJavaElement element) throws JavaModelException {
+        ISourceRange r = ((ISourceReference) element).getNameRange();
+        return JDTUtils.toRange(typeRoot, r.getOffset(), r.getLength());
+    }
+
+    static class MainMethod {
```

Comment:
Better utilize the org.eclipse.lsp4j if possible to make the design consistently.

Reply:
I saw there is a `CodeLens` type in lsp4j, it's suitable for Java language server, because all codelens operation (label, command) are implemented in Java code. But in Java debugger, we just take the Java code as an utility method, the action command is implemented in ts. If we put the codeLens command info in Java code, it looks like not clean.

Reply:
Then we should avoid use CodeLens in the java code for debugging, just maininfo or whatever. Otherwise, this will  mix concept in the code.

Reply:
Agree. Have changed the codeLens to mainMethod info.